src/Psalm/Internal/Analyzer/Statements/Expression/Fetch/ArrayFetchAnalyzer.php

```diff
@@ -2069,4 +2053,61 @@ private static function handleArrayAccessOnString(
             }
         }
     }
+
+    /**
+     * @param Atomic[] $offset_types
+     */
+    private static function checkArrayOffsetType(
+        Type\Union $offset_type,
+        array $offset_types,
+        \Psalm\Codebase $codebase
+    ): bool {
+        $has_valid_absolute_offset = false;
+        foreach ($offset_types as $atomic_offset_type) {
+            if ($atomic_offset_type instanceof Type\Atomic\TClassConstant) {
+                $expanded = TypeExpander::expandAtomic(
+                    $codebase,
+                    $atomic_offset_type,
+                    $atomic_offset_type->fq_classlike_name,
+                    $atomic_offset_type->fq_classlike_name,
+                    null,
+                    true,
+                    true
+                );
+
+                if ($expanded instanceof Atomic) {
+                    $has_valid_absolute_offset = self::checkArrayOffsetType($offset_type, [$expanded], $codebase);
+                } else {
+                    $has_valid_absolute_offset = self::checkArrayOffsetType($offset_type, $expanded, $codebase);
+                }
+
+                if ($has_valid_absolute_offset) {
+                    break;
+                }
+            }
+
+            if ($atomic_offset_type instanceof Type\Atomic\TFalse &&
+                $offset_type->ignore_falsable_issues === true
+            ) {
+                //do nothing
+            } elseif ($atomic_offset_type instanceof Type\Atomic\TNull &&
+                $offset_type->ignore_nullable_issues === true
+            ) {
+                //do nothing
+            } elseif ($atomic_offset_type instanceof Type\Atomic\TString ||
+                $atomic_offset_type instanceof Type\Atomic\TInt ||
+                $atomic_offset_type instanceof Type\Atomic\TArrayKey ||
+                $atomic_offset_type instanceof Type\Atomic\TMixed ||
+                $atomic_offset_type instanceof Type\Atomic\TTemplateParam ||
```

Comment:
Shouldn't `$atomic_offset_type->as` be checked here as well? E.g. `@template T of array` and using `T` as an offset is probably wrong.

Reply:
I'm adding for the sake of completeness, however, offset types are checked twice: 
- once against `array-key`
- once against the expected offset type (which could be anything because psalm will technically register a `array<array, string>`

The method you flagged is the first former. This means that you would need to have a source of `array<array, string>` somewhere to be able to give an array offset type to psalm without error.